lib/Sema/PCMacro.cpp

```diff
@@ -0,0 +1,718 @@
+          Added<Stmt *> LogBefore = buildLoggerCall(CS->getSourceRange(), true);
+          Added<Stmt *> LogAfter = buildLoggerCall(CS->getSourceRange(), false);
+          if (*LogBefore && *LogAfter) {
+            Elements[EI] = *LogBefore;
+            Elements.insert(Elements.begin() + (EI + 1), *LogAfter);
+            Elements.insert(Elements.begin() + (EI + 2), CS);
+            EI += 2;
+          }
+
+        } else if (BreakStmt *BS = dyn_cast<BreakStmt>(S)) {
+          Added<Stmt *> LogBefore = buildLoggerCall(BS->getSourceRange(), true);
+          Added<Stmt *> LogAfter = buildLoggerCall(BS->getSourceRange(), false);
+          if (*LogBefore && *LogAfter) {
+            Elements[EI] = *LogBefore;
+            Elements.insert(Elements.begin() + (EI + 1), *LogAfter);
+            Elements.insert(Elements.begin() + (EI + 2), BS);
+            EI += 2;
+          }
+
+        } else if (FallthroughStmt *FS = dyn_cast<FallthroughStmt>(S)) {
+          Added<Stmt *> LogBefore = buildLoggerCall(FS->getSourceRange(), true);
+          Added<Stmt *> LogAfter = buildLoggerCall(FS->getSourceRange(), false);
+          if (*LogBefore && *LogAfter) {
+            Elements[EI] = *LogBefore;
+            Elements.insert(Elements.begin() + (EI + 1), *LogAfter);
+            Elements.insert(Elements.begin() + (EI + 2), FS);
+            EI += 2;
+          }
+
+        } else {
+          Stmt *NS = transformStmt(S);
+          if (NS != S) {
+            Elements[EI] = NS;
+          }
+        }
+      } else if (Decl *D = Element.dyn_cast<Decl *>()) {
+        D->walk(CF);
+        if (auto *PBD = dyn_cast<PatternBindingDecl>(D)) {
+          // FIXME: Should iterate all var decls
+          if (VarDecl *VD = PBD->getSingleVar()) {
+            if (VD->getParentInitializer()) {
+
+              SourceRange SR = PBD->getSourceRange();
+              if (!SR.isValid()) {
+                SR = PBD->getOrigInitRange(0);
+              }
+
+              Added<Stmt *> LogBefore = buildLoggerCall(SR, true);
+              Added<Stmt *> LogAfter = buildLoggerCall(SR, false);
+
+              if (*LogBefore && *LogAfter) {
+                Elements[EI] = *LogBefore;
+                Elements.insert(Elements.begin() + (EI + 1), D);
+                Elements.insert(Elements.begin() + (EI + 2), *LogAfter);
+                EI += 2;
+              }
+            }
+          }
+        } else {
+          transformDecl(D);
+        }
+      }
+    }
+
+    return swift::BraceStmt::create(Context, BS->getLBraceLoc(), Elements,
+                                    BS->getRBraceLoc());
+  }
+
+  std::pair<PatternBindingDecl *, VarDecl *>
+  buildPatternAndVariable(Expr *InitExpr) {
+    char NameBuf[11] = {0};
+    snprintf(NameBuf, sizeof(NameBuf), "pctmp%u", TmpNameIndex);
+    TmpNameIndex++;
+
+    Expr *MaybeLoadInitExpr = nullptr;
+
+    if (LValueType *LVT =
+            dyn_cast<LValueType>(InitExpr->getType().getPointer())) {
```

Comment:
If you want to look through sugar, use `InitExpr->getType()->getAs<LValueType>` rather than this. (Most of the time that's what you want.)